### PR TITLE
Fix 16628: "undefined" exception when name of binding element in binding pattern is empty

### LINKED ISSUE #16628: Cannot read property 'pos' of undefined
**TypeScript Version:**  2.3.4

**Code**

```ts
const [,] = [1,2];
```

**Expected behavior:**
Compiles with compile error or without.

**Actual behavior:**
Compiler does not emit a compile error, but throws:
```
/Users/chrmarti/.nvm/versions/node/v8.0.0/lib/node_modules/typescript/lib/tsc.js
:53887
                throw e;
                ^

TypeError: Cannot read property 'pos' of undefined
    at visitVariableDeclarationList (/Users/chrmarti/.nvm/versions/node/v8.0.0/l
ib/node_modules/typescript/lib/tsc.js:44380:90)
    at visitJavaScript (/Users/chrmarti/.nvm/versions/node/v8.0.0/lib/node_modul
es/typescript/lib/tsc.js:43555:28)
    at visitor (/Users/chrmarti/.nvm/versions/node/v8.0.0/lib/node_modules/types
cript/lib/tsc.js:43516:24)
    at visitNode (/Users/chrmarti/.nvm/versions/node/v8.0.0/lib/node_modules/typ
escript/lib/tsc.js:39477:23)
    at Object.visitEachChild (/Users/chrmarti/.nvm/versions/node/v8.0.0/lib/node_modules/typescript/lib/tsc.js:39709:111)
    at visitVariableStatement (/Users/chrmarti/.nvm/versions/node/v8.0.0/lib/node_modules/typescript/lib/tsc.js:44358:30)
    at visitJavaScript (/Users/chrmarti/.nvm/versions/node/v8.0.0/lib/node_modules/typescript/lib/tsc.js:43625:28)
    at visitor (/Users/chrmarti/.nvm/versions/node/v8.0.0/lib/node_modules/typescript/lib/tsc.js:43516:24)
    at Object.visitNodes (/Users/chrmarti/.nvm/versions/node/v8.0.0/lib/node_modules/typescript/lib/tsc.js:39514:48)
    at visitSourceFile (/Users/chrmarti/.nvm/versions/node/v8.0.0/lib/node_modules/typescript/lib/tsc.js:43639:40)
```

### PR DESCRIPTION
**Pull request by yuit**
Fix #16628

### CODE REVIEW

File: src/compiler/declarationEmitter.ts [hunk 1]
```diff
@@ -1367,6 +1367,11 @@ namespace ts {
         }
 
         function writeVariableStatement(node: VariableStatement) {
+            // If binding pattern doesn't have name, then there is nothing to be emitted for declaration file i.e. const [,] = [1,2].
+            if (node.declarationList.declarations.length && isBindingPattern(node.declarationList.declarations[0].name) &&
```
Comment (weswigham) [inline src/compiler/declarationEmitter.ts]: This needs to be a recursive check, as binding patterns can nest infinitely, and of not just omitted expressions, but omitted expressions or binding patterns containing only the aforementioned.

Also, a test for that:
```ts
// @declaration: true
export let [,,[,[],,[],]] = undefined as any;
```

File: src/compiler/transformers/destructuring.ts [hunk 2]
```diff
@@ -331,11 +331,14 @@ namespace ts {
                 location
             );
         }
-        else if (numElements !== 1 && (flattenContext.level < FlattenLevel.ObjectRest || numElements === 0)) {
+        else if (numElements !== 1 && (flattenContext.level < FlattenLevel.ObjectRest || numElements === 0) ||
+                every(elements, (element) => isOmittedExpression(element))) {
```
Comment (weswigham) [inline src/compiler/transformers/destructuring.ts]: This... might? Need to be recursive? Depends on how the transform works.

File: src/compiler/transformers/es2015.ts [hunk 3]
```diff
@@ -2105,13 +2105,14 @@ namespace ts {
...
                 if (node.transformFlags & TransformFlags.ContainsBindingPattern
-                    && (isBindingPattern(node.declarations[0].name)
-                        || isBindingPattern(lastOrUndefined(node.declarations).name))) {
+                    && (isBindingPattern(node.declarations[0].name) || isBindingPattern(lastOrUndefined(node.declarations).name))) {
                     // If the first or last declaration is a binding pattern, we need to modify
                     // the source map range for the declaration list.
                     const firstDeclaration = firstOrUndefined(declarations);
                     const lastDeclaration = lastOrUndefined(declarations);
-                    setSourceMapRange(declarationList, createRange(firstDeclaration.pos, lastDeclaration.end));
+                    if (firstDeclaration && lastDeclaration) {
```
Comment (weswigham) [inline src/compiler/transformers/es2015.ts]: We should probably bail much before here - up before `createVariableDeclarationList` if `declarations` is empty.

File: src/compiler/transformers/destructuring.ts (hunk 2 above)
Comment (weswigham) [inline src/compiler/transformers/destructuring.ts]: Answer: No. To be spec compliant, we still emit temporaries in js for empty binding patterns in order to attempt to get the elements for the sub-binding patterns/invoke their side-effects (and potentially error if they do not exist).

File: src/compiler/transformers/es2015.ts [hunk 4]
```diff
@@ -2104,13 +2104,12 @@ namespace ts {
                 setTextRange(declarationList, node);
                 setCommentRange(declarationList, node);
 
-                if (node.transformFlags & TransformFlags.ContainsBindingPattern
-                    && (isBindingPattern(node.declarations[0].name)
-                        || isBindingPattern(lastOrUndefined(node.declarations).name))) {
+                const firstDeclaration = firstOrUndefined(declarations);
```
Comment (rbuckton) [inline src/compiler/transformers/es2015.ts]: Why look up the declarations before checking if there is work to do? If we need to ensure `firstDeclaration` is not `undefined`, add another `if` inside of the existing block.

Also, you shouldn't need to check _both_ `firstDeclaration` and `lastDeclaration` as both will be undefined if `declarations.length === 0` and both will have values if `declarations.length > 0`.

File: src/compiler/transformers/destructuring.ts [hunk 5]
```diff
@@ -331,11 +331,14 @@ namespace ts {
                 location
             );
         }
-        else if (numElements !== 1 && (flattenContext.level < FlattenLevel.ObjectRest || numElements === 0)) {
+        else if (numElements !== 1 && (flattenContext.level < FlattenLevel.ObjectRest || numElements === 0)
+            || every(elements, (element) => isOmittedExpression(element))) {
```
Comment (rbuckton) [inline src/compiler/transformers/destructuring.ts]: Can we just use `every(elements, isOmittedExpression)`? No need to introduce a new function here.

**Comment by weswigham** [comment]
@rbuckton Done

**Pull request merged**